proposals/js-string-builtins/Overview.md

```diff
@@ -28,221 +28,277 @@ The first problem is that existing API’s require a calling convention conversi
+/// outside the bounds of the array.
+///
+/// NOTE: This function only takes an immutable i16 array defined in its own
+/// recursion group.
+///
+/// If this is an issue for toolchains, we can look into how to relax the
+/// function type while still maintaining good performance.
+func fromCharCodeArray(
   array: (ref null (array i16)),
   start: i32,
   end: i32
 ) -> (ref extern)
 {
-  start = ToUint32(start);
-  end = ToUint32(end);
-
-  // [2]
-  if (end > array.length || start > end)
-    trap;
-
-  let result = "";
-  for(let i = start; i < end; i++) {
-    // [1], [4]
-    result += String.fromCharCode(array[i]);
-  }
-  return result;
-}
-```
-
-[4]: "array[i]" is meant to express "load the i-th element of the array", in Wasm terms: (array.get_u $i16-array-type (local.get $array) (local.get $i)) for an appropriate $i16-array-type.
-
-Note: This function takes an i16 array defined in its own recursion group. If this is an issue for a toolchain, we can look into how to relax the function type while still maintaining good performance.
+  // NOTE: `start` and `end` are interpreted as signed 32-bit integers when
+  // converted to JS values using standard conversions. Reinterpret them as
+  // unsigned here.
+  start >>>= 0;
+  end >>>= 0;
 
-### "wasm:js-string" "fromWtf8Array"
-
-```
-func fromWtf8Array(
-  array: (ref null (array i8)),
-  start: i32,
-  end: i32
-) -> (ref extern)
-{
-  start = ToUint32(start);
-  end = ToUint32(end);
+  if (array === null)
+    trap();
 
-  // [2]
-  if (end > array.length || start > end)
-    trap;
+  if (start > end ||
+      end > array_len(array))
+    trap();
 
-  // This summarizes as: "decode the WTF-8 string stored at array[start:end],
-  // or trap if there is no valid WTF-8 string there".
   let result = "";
-  while (start < end) {
-    if there is no valid wtf8-encoded code point at array[start]
-      trap;
-    let c be the code point at array[start];
-    // [1]
-    result += String.fromCodePoint(c);
-    increment start by as many bytes as it took to decode c;
+  for(let i = start; i < end; i++) {
+    let charCode = array_i16_get(array, i);
+    result += String.fromCharCode(charCode);
   }
   return result;
 }
 ```
 
-Note to implementers: while this is the only usage of WTF-8 in this document, it shouldn't be very burdensome to implement, because all existing strings in Wasm modules (import/export names, contents of the name section) are already in UTF-8 format, so implementations must already have decoding infrastructure for that. We need the relaxation from UTF-8 to WTF-8 to support WTF-16 based source languages, which may have unpaired surrogates in string constants in existing/legacy code.
-
-### "wasm:js-string" "toWtf16Array"
-
-"start" is the index in the array where the first codeunit of the string will be written.
-
-Returns the number of codeunits written. Traps if the string doesn't fit into the array.
+### "wasm:js-string" "copyToCharCodeArray"
```

Comment:
The reason is that it takes a pre-allocated array to copy into, instead of creating a new array. I figured that if we had support for a `string -> (array i8)` operation that creates an immutable array from a JS string, then we'd want to call that `fromCharCodeArray` and need some way to distinguish.

Reply:
How would `intoCharChodeArray` work for you? It matches the TextDecoder interface which has `to` and `into` versions for allocating/re-using existing buffers.

Reply:
Sure, that's fine. (It's also just a name, not very important either way.)

If we do want to unify names, `fromCharCodeArray` and `decodeStringFromUTF8Array` could also be made more similar to each other, but again it's just a name. (Some might argue "yeah, the names are different because one is a JS String operation and the other a TextDecoder thing", some might reply "what do I care? They're utf8 and wtf16 versions of the same operation, so they should have similar names".)

Reply:
Yeah, indeed my reasoning was 'the names are different because one is a JS String operation and the other a TextDecoder thing'. I think it's more compelling when you consider the full import specifier:

`"wasm:js-string" "fromCharCodeArray"`
1. `"wasm:text-decoder" "fromUTF8Array"` vs
2. `"wasm:text-decoder" "decodeStringFromUTF8Array"`

(1) appears to me to be creating a text decoder from a UTF-8 array.